controllers/space/space_controller.go

```diff
@@ -0,0 +1,343 @@
+			TemplateRef: tmpl.TemplateRef,
+		}
+	}
+
+	return nsTmplSet, nil
+}
+
+func (r *Reconciler) ensureSpaceDeletion(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	logger.Info("terminating Space")
+	if deleted, err := r.deleteNSTemplateSet(logger, space); err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, space, r.setStatusTerminatingFailed, err, "failed to delete the NSTemplateSet")
+	} else if deleted {
+		if err := r.setStatusTerminating(space); err != nil {
+			logger.Error(err, "error updating status")
+			return false, err
+		}
+		return true, nil // requeue until NSTemplateSet is fully deleted
+	}
+	// Remove finalizer from Space
+	util.RemoveFinalizer(space, toolchainv1alpha1.FinalizerName)
+	if err := r.Client.Update(context.TODO(), space); err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, space, r.setStatusTerminatingFailed, err, "failed to remove finalizer")
+	}
+	logger.Info("removed finalizer")
+	// no need to update the status of the Space once the finalizer has been removed, since
+	// the resource will be deleted
+	return false, nil // no need to requeue
+}
+
+// deleteNSTemplateSet triggers the deletion of the NSTemplateSet on the target member cluster.
+// Returns `true/nil` if the NSTemplateSet was deleted during this call, or if it is already being deleted
+// (ie, the deletion was triggered during a previous call/reconcile loop, and it's not fully done yet)
+// Returns `false/nil` if the NSTemplateSet was already deleted (ie, resource doesn't exist anymore),
+//   or if there is no target cluster specified in the given space, of if the target cluster is unknown.
+// Returns `false/error` if an error occurred
+func (r *Reconciler) deleteNSTemplateSet(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	if space.Spec.TargetCluster == "" {
+		logger.Info("cannot deleted NSTemplateSet: no target cluster specified")
+		return false, nil // abort deletion
+	}
+	memberCluster, found := r.MemberClusters[space.Spec.TargetCluster]
+	if !found {
+		logger.WithValues("target_cluster", space.Spec.TargetCluster).Info("Cannot deleted NSTemplateSet: unknown target member cluster")
+		return false, fmt.Errorf("unable to get the client for member cluster '%s'", space.Spec.TargetCluster)
+	}
+	// Get the NSTemplateSet associated with the Space
+	nstmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := memberCluster.Client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: memberCluster.OperatorNamespace,
+			Name:      space.Name},
+		nstmplSet)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err // something wrong happened
+		}
+		logger.Info("the NSTemplateSet resource is already deleted")
+		return false, nil // NSTemplateSet was already deleted
+	}
+	if util.IsBeingDeleted(nstmplSet) {
+		logger.Info("the NSTemplateSet resource is already being deleted")
+		return true, nil // requeue until fully deleted
+	}
```

Comment:
we should also check for how long the NStempalteSet is being deleted - to catch the cases when the deletion is blocked

Reply:
I suggest we address this in https://issues.redhat.com/browse/CRT-1393